File: js/lib/scraper.js
/**
 * Entra Portal Scraper
 * Handles all browser automation and data extraction from Entra portal
 */

const { chromium } = require("playwright");
const axios = require("axios");
const cheerio = require("cheerio");
const { waitForSplashScreen, clickTab, setDateRangeFilter, scrapeDetailsList } = require("./browser-helpers");
const { getConfig } = require("./config");

/**
 * Extract release type from the beginning of the title
 * @param {string} title - The full title string
 * @returns {{releaseType: string, cleanedTitle: string}}
 */
function extractReleaseTypeFromTitle(title) {
  const config = getConfig();
  const releaseTypeMapping = config.httpScraping.releaseTypeMapping;
  
  // Check if title starts with any known release type (case-insensitive)
  const titleLower = title.toLowerCase();
  for (const [pageValue, mappedValue] of Object.entries(releaseTypeMapping)) {
    if (titleLower.startsWith(pageValue.toLowerCase())) {
      // Extract the part after the release type
      let remainder = title.slice(pageValue.length).trim();
      // Remove common separators at the start
      const separators = ['-', ':', '–', '—'];
      for (const sep of separators) {
        if (remainder.startsWith(sep)) {
          remainder = remainder.slice(1).trim();
          break;
        }
      }
      return { releaseType: mappedValue, cleanedTitle: remainder };
    }
  }
  
  // No release type found - check if there's an unmapped one
  if ([' - ', ': ', ' – '].some(sep => title.slice(0, 50).includes(sep))) {
    for (const sep of [' - ', ': ', ' – ']) {
      if (title.includes(sep)) {
        const potentialType = title.split(sep)[0].trim();
        if (potentialType && /^[A-Z]/.test(potentialType)) {
          console.warn(`⚠️ Unmapped release type found: '${potentialType}' in title: ${title.slice(0, 60)}...`);
        }
        break;
      }
    }
  }
  
  return { releaseType: '', cleanedTitle: title };
}

/**
 * Extract a single item from the What's New page
 * @param {import('cheerio').Cheerio} h3Element - The h3 element
 * @param {string} monthText - The month/year text
 * @param {import('cheerio').CheerioAPI} $ - Cheerio instance
 * @returns {Object|null}
 */
function extractWhatsNewItem(h3Element, monthText, $) {
  try {
    const item = {
      releaseType: '',
      title: '',
      type: '',
      serviceCategory: '',
      productCapability: '',
      detail: '',
      link: '',
      date: ''
    };
    
    // Extract title and link
    const titleLink = h3Element.find('a');
    let fullTitle;
    if (titleLink.length > 0) {
      fullTitle = titleLink.text().trim();
      item.link = titleLink.attr('href') || '';
      if (item.link && !item.link.startsWith('http')) {
        const pageUrl = getConfig().httpScraping.whatsNew;
        item.link = new URL(item.link, pageUrl).href;
      }
    } else {
      fullTitle = h3Element.text().trim();
    }
    
    // Extract release type from title
    const { releaseType, cleanedTitle } = extractReleaseTypeFromTitle(fullTitle);
    
    if (!releaseType && [' - ', ': ', ' – '].some(sep => fullTitle.includes(sep))) {
      for (const sep of [' - ', ': ', ' – ']) {
        if (fullTitle.includes(sep)) {
          const potentialType = fullTitle.split(sep)[0].trim();
          if (potentialType && /^[A-Z]/.test(potentialType)) {
            console.warn(`⚠️ Unmapped release type found: '${potentialType}' in title: ${fullTitle.slice(0, 60)}...`);
          }
          break;
        }
      }
    }
    
    item.releaseType = releaseType;
    item.title = cleanedTitle;
    item.date = monthText;
    
    // Extract detail from following paragraphs
    const detailParts = [];
    let current = h3Element.next();
    let lastBodyLinkHref = '';
    
    while (current.length > 0 && !['h2', 'h3'].includes(current.prop('tagName')?.toLowerCase())) {
      if (current.prop('tagName')?.toLowerCase() === 'p') {
        const text = current.text().trim();
        if (text) {
          // Look for metadata in strong tags
          current.find('strong').each((_, strong) => {
            const label = $(strong).text().trim().replace(/:$/, '');
            const nextText = $(strong)[0]?.nextSibling;
            if (nextText && nextText.type === 'text') {
              const value = nextText.data.trim().replace(/^:/, '');
              
              if (label.includes('Type') && !item.type) {
                item.type = value;
              } else if (label.toLowerCase().includes('service category')) {
                item.serviceCategory = value;
              } else if (label.toLowerCase().includes('product capability')) {
                item.productCapability = value;
              }
            }
          });
          
          detailParts.push(text);
        }
      } else if (current.prop('tagName')?.toLowerCase() === 'ul') {
        current.find('li').each((_, li) => {
          detailParts.push(`• ${$(li).text().trim()}`);
        });
      }
      
      // Track the last hyperlink in the body as a fallback for entries whose
      // title has no link (the reference link's preceding text may vary).
      const bodyLinks = current.find('a[href]');
      if (bodyLinks.length > 0) {
        lastBodyLinkHref = bodyLinks.last().attr('href') || lastBodyLinkHref;
      }
      
      current = current.next();
    }
    
    item.detail = detailParts.join(' ');
    
    // Fallback: if the title had no anchor, use the last link found in the body
    // (e.g. "For more information, see: <link>"). Resolve relative hrefs against
    // the page URL so paths like "../identity/..." expand correctly.
    if (!item.link && lastBodyLinkHref) {
      item.link = new URL(lastBodyLinkHref, getConfig().httpScraping.whatsNew).href;
    }
    
    return item.title ? item : null;
  } catch (error) {
    console.warn(`⚠️ Error extracting item: ${error.message}`);
    return null;
  }
}

/**
 * Scrape data from Microsoft Learn What's New page
 * @returns {Promise<Array<Object>|null>}
 */
async function scrapeWhatsNewPage() {
  const config = getConfig();
  const url = config.httpScraping.whatsNew;
  
  try {
    console.log(`🌐 Fetching ${url}...`);
    const response = await axios.get(url, { timeout: 30000 });
    
    const $ = cheerio.load(response.data);
    const items = [];
    
    const months = ['January', 'February', 'March', 'April', 'May', 'June', 
                    'July', 'August', 'September', 'October', 'November', 'December'];
    
    // Find all h2 headers that represent month sections
    $('h2[id]').each((_, monthSection) => {
      const monthText = $(monthSection).text().trim();
      
      // Skip if not a date header
      if (!months.some(month => monthText.includes(month))) {
        return;
      }
      
      // Find all h3 items under this month
      let current = $(monthSection).next();
      while (current.length > 0 && current.prop('tagName')?.toLowerCase() !== 'h2') {
        if (current.prop('tagName')?.toLowerCase() === 'h3') {
          const item = extractWhatsNewItem(current, monthText, $);
          if (item) {
            items.push(item);
          }
        }
        current = current.next();
      }
    });
    
    console.log(`✅ Extracted ${items.length} items from What's New page`);
    return items;
  } catch (error) {
    console.error(`❌ Error scraping What's New page: ${error.message}`);
    return null;
  }
}

/**
 * Initialize browser and navigate to Entra portal
 * @returns {Promise<{context: import('playwright').BrowserContext, page: import('playwright').Page, frame: import('playwright').Frame}>}
 */
async function initializeBrowser() {
  const config = getConfig();
  const headless = config?.browserScraping?.headless ?? false;

  const context = await chromium.launchPersistentContext('./edge-profile', {
    channel: 'msedge',
    headless,
    args: ['--disable-blink-features=AutomationControlled'],
    viewport: { width: 1920, height: 1080 },
    deviceScaleFactor: 1,
  });
  
  const page = context.pages()[0];

  const entraUrl = config.browserScraping.entraPortal;

  // Navigate to Entra portal
  await page.goto(entraUrl, { waitUntil: 'domcontentloaded', timeout: 60000 });

  // Get the main iframe
  const iframeLocator = page.locator('iframe[name="ChangeManagementHubList.ReactView"]');
  await iframeLocator.waitFor({ state: 'attached', timeout: 30000 });

  const iframeHandle = await iframeLocator.elementHandle();
  const frame = await iframeHandle.contentFrame();
  
  if (!frame) {
    throw new Error('ReactView frame attached, but content not available yet.');
  }

  // Wait for initial splash screen to disappear
  await waitForSplashScreen(frame, 60000);

  // Wait for progress dots to disappear
  try {
    const progressDots = frame.locator('div.fxs-progress-dots');
    await progressDots.waitFor({ state: 'hidden', timeout: 15000 });
  } catch (err) {
    // Progress dots might not exist or already hidden
  }

  return { context, page, frame };
}

/**
 * Scrape data from a specific tab
 * @param {import('playwright').Page} page - The page
 * @param {import('playwright').Frame} frame - The iframe
 * @param {string|RegExp} tabName - The name of the tab to scrape
 * @param {string|null} dateFilter - Optional date filter to apply
 * @param {boolean} extractDetails - Whether to extract details by clicking rows
 * @returns {Promise<Array<Object>|null>} Scraped data or null if tab not found
 */
async function scrapeTab(page, frame, tabName, dateFilter = null, extractDetails = true) {
  // Click the tab
  const tabClicked = await clickTab(frame, tabName);

  if (!tabClicked) {
    console.error(`❌ Could not locate the ${tabName} tab/menu.`);
    return null;
  }

  // Set date filter if specified
  if (dateFilter) {
    const filterSet = await setDateRangeFilter(frame, dateFilter);
    if (!filterSet) {
      console.warn('⚠️ Could not set date range filter, continuing anyway...');
    }
  }

  // Scrape the data
  const data = await scrapeDetailsList(page, frame, extractDetails);
  console.log(`✅ Extracted ${data.length} items from ${tabName}`);

  return data;
}

/**
 * Scrape Roadmap data from Entra portal
 * @returns {Promise<{roadmap: Array<Object>|null, changeAnnouncements: Array<Object>|null}>}
 */
async function scrapeEntraPortal() {
  let context, page, frame;
  
  try {
    // Initialize browser and navigate
    ({ context, page, frame } = await initializeBrowser());

    const config = getConfig();
    const dateFilter = config.browserScraping.dateFilter;

    // Scrape Roadmap
    const roadmapExtractDetails = config.browserScraping.roadmap.extractDetails;
    const roadmap = await scrapeTab(page, frame, /^Roadmap$/i, dateFilter, roadmapExtractDetails);

    // Scrape Change Announcements
    const changeAnnouncementsExtractDetails = config.browserScraping.changeAnnouncements.extractDetails;
    const changeAnnouncements = await scrapeTab(page, frame, /^Change announcements$/i, dateFilter, changeAnnouncementsExtractDetails);

    return { roadmap, changeAnnouncements };
  } finally {
    if (context) {
      await context.close();
      console.log('✅ Browser closed.');
    }
  }
}

/**
 * Scrape data from all sources: Entra portal and Microsoft Learn What's New
 * @returns {Promise<{roadmap: Array<Object>|null, changeAnnouncements: Array<Object>|null, whatsNew: Array<Object>|null}>}
 */
async function scrapeAllSources() {
  // Scrape Entra portal data
  const portalData = await scrapeEntraPortal();
  
  // Scrape What's New page
  console.log("\n📚 Scraping Microsoft Learn What's New page...");
  const whatsNew = await scrapeWhatsNewPage();
  
  return {
    roadmap: portalData.roadmap,
    changeAnnouncements: portalData.changeAnnouncements,
    whatsNew
  };
}

module.exports = {
  initializeBrowser,
  scrapeTab,
  scrapeEntraPortal,
  scrapeAllSources,
};

File: python/lib/scraper.py
"""
Entra Portal Scraper
Handles all browser automation and data extraction from Entra portal
"""

import re
from pathlib import Path
from typing import Dict, Any, List, Optional
from datetime import datetime

from playwright.async_api import async_playwright, Page, Frame
from bs4 import BeautifulSoup
import requests

from .browser_helpers import (
    wait_for_splash_screen,
    click_tab,
    set_date_range_filter,
    scrape_details_list
)
from .auth import get_configuration


def extract_release_type_from_title(title: str) -> tuple[str, str]:
    """
    Extract release type from the beginning of the title.
    
    Args:
        title: The full title string
        
    Returns:
        Tuple of (release_type, cleaned_title)
    """
    config = get_configuration().get('config', {})
    release_type_mapping = config.get('releaseTypeMapping', {})
    
    # Check if title starts with any known release type (case-insensitive)
    title_lower = title.lower()
    for page_value, mapped_value in release_type_mapping.items():
        if title_lower.startswith(page_value.lower()):
            # Extract the part after the release type
            remainder = title[len(page_value):].strip()
            # Remove common separators at the start
            for sep in ['-', ':', '–', '—']:
                if remainder.startswith(sep):
                    remainder = remainder[1:].strip()
                    break
            return mapped_value, remainder
    
    # No release type found - check if there's an unmapped one
    if any(sep in title[:50] for sep in [' - ', ': ', ' – ']):
        for sep in [' - ', ': ', ' – ']:
            if sep in title:
                potential_type = title.split(sep)[0].strip()
                if potential_type and potential_type[0].isupper():
                    print(f"⚠️ Unmapped release type found: '{potential_type}' in title: {title[:60]}...")
                break
    
    return '', title


async def scrape_tab(
    page: Page,
    frame: Frame,
    tab_name: str,
    date_filter: Optional[str] = None,
    extract_details: bool = True
) -> Optional[List[Dict[str, Any]]]:
    """
    Scrape data from a specific tab.
    
    Args:
        page: The page
        frame: The iframe
        tab_name: The name of the tab to scrape
        date_filter: Optional date filter to apply
        extract_details: Whether to extract details by clicking rows
        
    Returns:
        Scraped data or None if tab not found
    """
    # Click the tab
    tab_clicked = await click_tab(frame, tab_name)
    
    if not tab_clicked:
        print(f"❌ Could not locate the {tab_name} tab/menu.")
        return None
    
    # Set date filter if specified
    if date_filter:
        filter_set = await set_date_range_filter(frame, date_filter)
        if not filter_set:
            print('⚠️ Could not set date range filter, continuing anyway...')
    
    # Scrape the data
    data = await scrape_details_list(page, frame, extract_details)
    print(f"✅ Extracted {len(data)} items from {tab_name}")
    
    return data


def scrape_whats_new_page() -> Optional[List[Dict[str, Any]]]:
    """
    Scrape data from Microsoft Learn What's New page.
    
    Returns:
        List of items with Release Type, Title, Type, Service Category, 
        Product Capability, Detail, Link, Date
    """
    config = get_configuration().get('config', {})
    url = config.get('urls', {}).get('whatsNew', "https://learn.microsoft.com/en-us/entra/fundamentals/whats-new")
    
    try:
        print(f"🌐 Fetching {url}...")
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.content, 'html.parser')
        items = []
        
        # Find all h2 headers that represent month sections
        month_sections = soup.find_all('h2', id=True)
        
        for month_section in month_sections:
            month_text = month_section.get_text(strip=True)
            
            # Skip if not a date header (e.g., "January 2026")
            if not any(month in month_text for month in ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']):
                continue
            
            # Find all h3 items under this month
            current = month_section.find_next_sibling()
            while current and current.name != 'h2':
                if current.name == 'h3':
                    item = extract_whats_new_item(current, month_text)
                    if item:
                        items.append(item)
                current = current.find_next_sibling()
        
        print(f"✅ Extracted {len(items)} items from What's New page")
        return items
        
    except Exception as e:
        print(f"❌ Error scraping What's New page: {e}")
        return None


def extract_whats_new_item(h3_element, month_text: str) -> Optional[Dict[str, Any]]:
    """
    Extract a single item from the What's New page.
    
    Args:
        h3_element: The h3 element containing the item title
        month_text: The month/year text (e.g., "January 2026")
        
    Returns:
        Dictionary with item data
    """
    try:
        item = {
            'releaseType': '',
            'title': '',
            'type': '',
            'serviceCategory': '',
            'productCapability': '',
            'detail': '',
            'link': '',
            'date': ''
        }
        
        # Extract title and link
        title_link = h3_element.find('a')
        if title_link:
            full_title = title_link.get_text(strip=True)
            item['link'] = title_link.get('href', '')
            if item['link'] and not item['link'].startswith('http'):
                config = get_configuration().get('config', {})
                base_url = config.get('urls', {}).get('microsoftLearnBase', 'https://learn.microsoft.com')
                item['link'] = f"{base_url}{item['link']}"
        else:
            full_title = h3_element.get_text(strip=True)
        
        # Extract release type from title
        release_type, cleaned_title = extract_release_type_from_title(full_title)
        
        if not release_type and any(sep in full_title for sep in [' - ', ': ', ' – ']):
            # There might be an unmapped release type
            for sep in [' - ', ': ', ' – ']:
                if sep in full_title:
                    potential_type = full_title.split(sep)[0].strip()
                    # Check if it looks like a release type (title case, multiple words)
                    if potential_type and potential_type[0].isupper():
                        print(f"⚠️ Unmapped release type found: '{potential_type}' in title: {full_title[:60]}...")
                    break
        
        item['releaseType'] = release_type
        item['title'] = cleaned_title
        
        # Set date from month section
        item['date'] = month_text
        
        # Extract detail from the following paragraph(s)
        detail_parts = []
        current = h3_element.find_next_sibling()
        
        while current and current.name not in ['h2', 'h3']:
            if current.name == 'p':
                text = current.get_text(strip=True)
                if text:
                    # Look for metadata in strong tags
                    strong_tags = current.find_all('strong')
                    for strong in strong_tags:
                        label = strong.get_text(strip=True).rstrip(':')
                        # Get the text after the strong tag
                        next_text = strong.next_sibling
                        if next_text and isinstance(next_text, str):
                            value = next_text.strip().strip(':')
                            
                            if 'Type' in label and not item['type']:
                                item['type'] = value
                            elif 'Service category' in label or 'Service Category' in label:
                                item['serviceCategory'] = value
                            elif 'Product capability' in label or 'Product Capability' in label:
                                item['productCapability'] = value
                            elif 'Release' in label:
                                item['releaseType'] = value
                    
                    detail_parts.append(text)
            elif current.name == 'ul':
                # Add list items to detail
                list_items = current.find_all('li')
                for li in list_items:
                    detail_parts.append(f"• {li.get_text(strip=True)}")
            
            current = current.find_next_sibling()
        
        item['detail'] = ' '.join(detail_parts)
        
        # Only return item if it has a title
        return item if item['title'] else None
        
    except Exception as e:
        print(f"⚠️ Error extracting item: {e}")
        return None


async def scrape_entra_portal(
    date_filter: Optional[str] = None
) -> Dict[str, Optional[List[Dict[str, Any]]]]:
    """
    Scrape Roadmap data from Entra portal.
    
    Args:
        date_filter: Optional date filter
        
    Returns:
        Dictionary with roadmap and changeAnnouncements data
    """
    config = get_configuration().get('config', {})
    entra_url = config.get('urls', {}).get('entraPortal', "https://entra.microsoft.com/#blade/Microsoft_AAD_IAM/ChangeManagementHubList.ReactView")
    
    async with async_playwright() as playwright:
        profile_dir = Path(__file__).resolve().parent.parent / "edge-profile"
        edge_exe = Path("C:/Program Files/Microsoft/Edge/Application/msedge.exe")

        launch_kwargs = {
            "headless": False,
            "args": ["--disable-blink-features=AutomationControlled"],
            "viewport": {"width": 1920, "height": 1080},
            "device_scale_factor": 1,
        }

        if edge_exe.exists():
            launch_kwargs["executable_path"] = str(edge_exe)
        else:
            launch_kwargs["channel"] = "msedge"

        context = await playwright.chromium.launch_persistent_context(
            str(profile_dir),
            **launch_kwargs,
        )

        try:
            page = context.pages[0] if context.pages else await context.new_page()

            # Navigate to Entra portal
            await page.goto(entra_url, wait_until='domcontentloaded', timeout=60000)

            # Get the main iframe
            iframe_locator = page.locator('iframe[name="ChangeManagementHubList.ReactView"]')
            await iframe_locator.wait_for(state='attached', timeout=30000)

            iframe_handle = await iframe_locator.element_handle()
            frame = await iframe_handle.content_frame()

            if not frame:
                raise RuntimeError('ReactView frame attached, but content not available yet.')

            # Wait for initial splash screen to disappear
            await wait_for_splash_screen(frame, 60000)

            # Wait for progress dots to disappear
            try:
                progress_dots = frame.locator('div.fxs-progress-dots')
                await progress_dots.wait_for(state='hidden', timeout=15000)
            except Exception:
                # Progress dots might not exist or already hidden
                pass

            # Get config for extract_details settings
            config_obj = get_configuration().get('config', {})
            lists_config = config_obj.get('lists', {})

            # Scrape Roadmap
            roadmap_extract_details = lists_config.get('roadmap', {}).get('extractDetails', True)
            roadmap = await scrape_tab(page, frame, '/^Roadmap$/i', date_filter, roadmap_extract_details)

            # Scrape Change Announcements
            change_announcements_extract_details = lists_config.get('changeAnnouncements', {}).get('extractDetails', True)
            change_announcements = await scrape_tab(page, frame, '/^Change announcements$/i', date_filter, change_announcements_extract_details)

            return {
                'roadmap': roadmap,
                'changeAnnouncements': change_announcements
            }
        finally:
            await context.close()
            print('✅ Browser closed.')


async def scrape_all_sources(
    date_filter: Optional[str] = None
) -> Dict[str, Optional[List[Dict[str, Any]]]]:
    """
    Scrape data from all sources: Entra portal and Microsoft Learn What's New.
    
    Args:
        date_filter: Optional date filter for portal scraping
        
    Returns:
        Dictionary with roadmap, changeAnnouncements, and whatsNew data
    """
    # Scrape Entra portal data
    portal_data = await scrape_entra_portal(date_filter)
    
    # Scrape What's New page
    print("\n📚 Scraping Microsoft Learn What's New page...")
    whats_new = scrape_whats_new_page()
    
    return {
        'roadmap': portal_data.get('roadmap'),
        'changeAnnouncements': portal_data.get('changeAnnouncements'),
        'whatsNew': whats_new
    }
